### PR TITLE
Enabling / Disabling semi-sync on ChangeType.

### PR DESCRIPTION
**Pull request by alainjobart**
Should fix:
https://github.com/youtube/vitess/issues/2195

### CODE REVIEW

File: go/vt/tabletmanager/state_change.go [hunk 1]
```diff
@@ -293,6 +293,27 @@ func (agent *ActionAgent) changeCallback(ctx context.Context, oldTablet, newTabl
...
+		isMasterEligible := isMasterEligible(newTablet.Type)
+		if wasMasterEligible && !isMasterEligible {
+			if err := agent.MysqlDaemon.SetSemiSyncEnabled(false, false); err != nil {
+				log.Errorf("Failed to SetSemiSyncEnabled(false, false), may not ack correctly: %v", err)
+			}
+		}
+		if !wasMasterEligible && isMasterEligible && newTablet.Type != topodatapb.TabletType_MASTER {
+			// We go from any type to REPLICA, let's
+			// enable semi-sync acks.
+			if err := agent.MysqlDaemon.SetSemiSyncEnabled(false, true); err != nil {
```
Comment (pivanof) [inline go/vt/tabletmanager/state_change.go]: Note that MySQL won't actually start sending semi-sync acks here unless you stop slave and then start it again. And I believe #2195 is exactly about that, i.e. rpl_semi_sync_slave_enabled was somehow set already, but there was no StopSlave+StartSlave for it to start working.

File: go/vt/tabletmanager/state_change.go [hunk 2]
```diff
@@ -293,6 +293,27 @@ func (agent *ActionAgent) changeCallback(ctx context.Context, oldTablet, newTabl
 		}
 	}
 
+	// See if we need to enable or disable semi-sync replication.
+	// Note we never enable the master side, as that would have
+	// been done by either the reparent process, or an external
+	// tool (like Orchestrator).
+	if *enableSemiSync {
```
Comment (pivanof) [inline go/vt/tabletmanager/state_change.go]: Can all this logic be combined with ActionAgent.enableSemiSync?

Comment (alainjobart) [inline go/vt/tabletmanager/state_change.go]: I don't think so, ActionAgent.enableSemiSync is for the new state only, this is for a transition.

File: go/vt/tabletmanager/state_change.go (hunk 1 above)
Comment (alainjobart) [inline go/vt/tabletmanager/state_change.go]: I uploaded a new commit to fix this, PTAL.

File: go/vt/tabletmanager/state_change.go (hunk 2 above)
Comment (pivanof) [inline go/vt/tabletmanager/state_change.go]: So what's preventing us from making ActionAgent.enableSemiSync to be for all transitions? And actually here in this case we don't really have to verify whether we actually transitioned or not, we can just set/reset semi-sync to be compliant with what we want, right? If it won't change anything and will just stop/start slave, that won't hurt.

Comment (alainjobart) [inline go/vt/tabletmanager/state_change.go]: I am looking at this right now. I think I found one bug at least: after a backup restore, it seems we use the wrong type (RESTORE) to decide if we need to enable semi-sync.

So I think I will change the logic of the checker part: as you suggest, always call enableSemiSync after a state change. But then, change the code: if replication is running, and the semi-sync status is different than the variable, restart replication.

**Comment by alainjobart** [comment]
@pivanof PTAL I ended up refactoring things a bit, I think this is better, and more aligned with what we want.

File: go/vt/tabletmanager/rpc_replication.go [hunk 3]
```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
 	return false
 }
 
-func (agent *ActionAgent) enableSemiSync(master bool) error {
+func (agent *ActionAgent) setSemiSync(tabletType topodatapb.TabletType) error {
```
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: I'd recommend a different method name. "setSemiSync" suggests setting it to while you may actually reset it. Maybe "adjustSemiSync", or "fixSemiSync"?

File: go/vt/tabletmanager/rpc_replication.go [hunk 4]
```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
...
+	// restart replication.
+	if _, err := agent.MysqlDaemon.SlaveStatus(); err != nil {
+		// Replication is not running, nothing to do.
+		return nil
+	}
+
+	shouldAck := isMasterEligible(tabletType)
+	acking, err := agent.MysqlDaemon.SemiSyncSlaveStatus()
+	if err != nil {
+		return fmt.Errorf("failed to get SemiSyncSlaveStatus: %v", err)
```
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: Should we just try to proceed in this case as if shouldAck != acking?
And actually why do we try to verify the status, why are we afraid to execute StopSlave/StartSlave when it's not necessary? I don't see any harm in that.

File: test/backup.py [hunk 5]
```diff
@@ -123,19 +123,29 @@ def _check_data(self, t, count, msg):
...
+    else:
+      t.check_db_var('rpl_semi_sync_slave_enabled', 'OFF')
+      t.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+
   def _reset_tablet_dir(self, t):
     """Stop mysql, delete everything including tablet dir, restart mysql."""
     utils.wait_procs([t.teardown_mysql()])
-    t.remove_tree()
+    # specify ignore_options because we want to delete the tree even
+    # if -k was specified on the command line.
```
Comment (pivanof) [inline test/backup.py]: Can you explain what -k on the command line means? And is it command line to the test?

File: test/backup.py [hunk 6]
```diff
@@ -168,6 +184,9 @@ def test_backup(self):
     - bring up tablet_replica2 after the fact, let it restore the backup
     - check all data is right (before+after backup data)
     - list the backup, remove it
+
+    Args:
+      tablet_type: replica or rdonly.
```
Comment (pivanof) [inline test/backup.py]: Should probably be "'replica' or 'rdonly'", i.e. values surrounded with quotes, otherwise it's not clear what these words mean.

File: test/reparent.py [hunk 7]
```diff
@@ -621,6 +621,60 @@ def test_reparent_with_down_slave(self, shard_id='0'):
...
+  def test_change_type_semi_sync(self):
+    utils.run_vtctl(['CreateKeyspace', 'test_keyspace'])
+
+    # create the database so vttablets start, as they are serving
+    tablet_62344.create_db('vt_test_keyspace')
+    tablet_62044.create_db('vt_test_keyspace')
+    tablet_41983.create_db('vt_test_keyspace')
+    tablet_31981.create_db('vt_test_keyspace')
+
+    # Start up a soon-to-be master, 3 replicas mysql and vttablet
```
Comment (pivanof) [inline test/reparent.py]: I couldn't quite understand this comment. Did you mean "Start up a soon-to-be master, 2 replicas and rdonly"?

File: test/reparent.py [hunk 8]
```diff
@@ -621,6 +621,60 @@ def test_reparent_with_down_slave(self, shard_id='0'):
...
+    # Force the slaves to reparent assuming that all the datasets are
+    # identical.
+    utils.run_vtctl(['InitShardMaster', '-force', 'test_keyspace/0',
+                     tablet_62344.tablet_alias], auto_log=True)
+    utils.validate_topology(ping_tablets=True)
+    self._check_master_tablet(tablet_62344)
+
+    # Stop replication on one slave, to make sure when we make it
+    # rdonly it doesn't start.
+    utils.run_vtctl(['StopSlave', tablet_41983.tablet_alias])
```
Comment (pivanof) [inline test/reparent.py]: This part doesn't seem to be related since you don't use tablet_41983 after this and don't make it rdonly.

File: go/vt/tabletmanager/state_change.go [hunk 9]
```diff
@@ -272,19 +272,19 @@ func (agent *ActionAgent) changeCallback(ctx context.Context, oldTablet, newTabl
 		}
 	}
 
-	// update stream needs to be started or stopped too
+	// UpdateStream needs to be started or stopped too.
```
Comment (pivanof) [inline go/vt/tabletmanager/state_change.go]: These changes are now completely unrelated. :-(

File: go/vt/tabletmanager/rpc_replication.go [hunk 10]
```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
...
+		return agent.MysqlDaemon.SetSemiSyncEnabled(false, false)
 	}
 
 	// Always enable slave-side since it doesn't hurt to keep it on for a master.
 	// The master-side needs to be off for a slave, or else it will get stuck.
-	return agent.MysqlDaemon.SetSemiSyncEnabled(master, true)
+	return agent.MysqlDaemon.SetSemiSyncEnabled(tabletType == topodatapb.TabletType_MASTER, true)
+}
+
+func (agent *ActionAgent) setSemiSyncAndFixReplication(tabletType topodatapb.TabletType) error {
```
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: If you name the method above "fixSemiSync" then this could be very conveniently named "fixSemiSyncAndReplication". It would be much better.

File: go/vt/tabletmanager/rpc_replication.go [hunk 11]
```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
...
+	if shouldAck == acking {
+		return nil
+	}
+
+	// We need to restart replication
+	log.Infof("Restarting replication for semi-sync flag change to take effect from %v to %v", acking, shouldAck)
+	if err := mysqlctl.StopSlave(agent.MysqlDaemon, agent.hookExtraEnv()); err != nil {
+		return fmt.Errorf("failed to StopSlave: %v", err)
+	}
+	if err := mysqlctl.StartSlave(agent.MysqlDaemon, agent.hookExtraEnv()); err != nil {
```
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: BTW, should we account here for a case when someone executed StopSlave first and then ChangeType, i.e. they don't want StartSlave to be automatically executed on any ChangeType? Besides if I'm not mistaken in such case StopSlave above will fail because the slave is already stopped.

File: go/vt/tabletmanager/rpc_replication.go (hunk 3 above)
Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: Done.

File: go/vt/tabletmanager/rpc_replication.go (hunk 10 above)
Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: Done.

File: go/vt/tabletmanager/rpc_replication.go (hunk 4 above)
Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: Next I also want to call this function in more places, like in RefreshState. So with this logic, it's better, we don't stop/start replication when we don't need to.

With the error case, I'd rather stop asap. If we can't get the flag, it means we can't reliably connect to mysql, I'd rather not touch it more.

File: go/vt/tabletmanager/rpc_replication.go (hunk 11 above)
Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: We already handle the replication not running case above, that's why we do SlaveStatus(). So by the time we get here, replication is running,

File: go/vt/tabletmanager/state_change.go (hunk 9 above)
Comment (alainjobart) [inline go/vt/tabletmanager/state_change.go]: Yes they are.

File: test/backup.py (hunk 5 above)
Comment (alainjobart) [inline test/backup.py]: Done.

File: test/backup.py (hunk 6 above)
Comment (alainjobart) [inline test/backup.py]: Good point, done.

File: test/reparent.py (hunk 7 above)
Comment (alainjobart) [inline test/reparent.py]: Yes, fixed.

File: test/reparent.py (hunk 8 above)
Comment (alainjobart) [inline test/reparent.py]: Oops I removed that part by mistake, re-added that test.

File: go/vt/tabletmanager/state_change.go (hunk 9 above)
Comment (pivanof) [inline go/vt/tabletmanager/state_change.go]: How they are related?

File: go/vt/tabletmanager/rpc_replication.go [hunk 12]
```diff
@@ -555,14 +547,63 @@ func isMasterEligible(tabletType topodatapb.TabletType) bool {
...
+		return agent.MysqlDaemon.SetSemiSyncEnabled(false, false)
 	}
 
 	// Always enable slave-side since it doesn't hurt to keep it on for a master.
 	// The master-side needs to be off for a slave, or else it will get stuck.
-	return agent.MysqlDaemon.SetSemiSyncEnabled(master, true)
+	return agent.MysqlDaemon.SetSemiSyncEnabled(tabletType == topodatapb.TabletType_MASTER, true)
+}
+
+func (agent *ActionAgent) fixSemiSyncAndFixReplication(tabletType topodatapb.TabletType) error {
```
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: Do you like two words "fix" in the function name? I was proposing to have only one.

File: go/vt/tabletmanager/rpc_replication.go (hunk 11 above)
Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: Just writing down the result of our discussion:
- the slave status didn't return, but the SyncStatus returned false, and shouldAck is false as well, so we didn't do anything.
- the test would have failed for a rdonly -> replica transition, as shouldAck would be true.

So I added a test case for that, saw it fail, and then also looked at the IO thread status. Now it should be correct.

File: go/vt/tabletmanager/state_change.go (hunk 9 above)
Comment (alainjobart) [inline go/vt/tabletmanager/state_change.go]: I meant yes, they are unrelated. They're tiny though, so I'll leave them in there.

**Comment by alainjobart** [comment]
@pivanof PTAL had a to fix more things for tests to pass.

File: go/vt/tabletmanager/rpc_replication.go [hunk 13]
```diff
@@ -254,6 +260,19 @@ func (agent *ActionAgent) InitSlave(ctx context.Context, parent *topodatapb.Tabl
 	}
 	agent.initReplication = true
 
+	// If we were a master type, switch our type to replica.  This
+	// is used on the old master when using InitShardMaster with
+	// -force, and the new master is different from the old master.
+	if agent.Tablet().Type == topodatapb.TabletType_MASTER {
+		if _, err := topotools.ChangeType(ctx, agent.TopoServer, agent.TabletAlias, topodatapb.TabletType_REPLICA); err != nil {
```
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: This will already invoke fixSemiSync, right? So maybe we should move call to fixSemiSync from above to "else" of this "if" statement?

File: test/reparent.py [hunk 14]
```diff
@@ -621,6 +621,86 @@ def test_reparent_with_down_slave(self, shard_id='0'):
...
+    tablet_31981.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+
+    # Change replica to rdonly while replicating, should turn off semi-sync,
+    # and restart replication.
+    utils.run_vtctl(['ChangeSlaveType', tablet_62044.tablet_alias, 'rdonly'],
+                    auto_log=True)
+    tablet_62044.check_db_var('rpl_semi_sync_slave_enabled', 'OFF')
+    tablet_62044.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+
+    # Change rdonly1 to replica, should turn on semi-sync, and not start rep.
```
Comment (pivanof) [inline test/reparent.py]: This test is very hard to follow now with all these numerically-named tablets. It would be great if they had better aliases, like tablet_replica, tablet_rdonly1, tablet_rdonly2, etc.

File: test/reparent.py [hunk 15]
```diff
@@ -621,6 +621,86 @@ def test_reparent_with_down_slave(self, shard_id='0'):
...
+    utils.run_vtctl(['ChangeSlaveType', tablet_41983.tablet_alias, 'replica'],
+                    auto_log=True)
+    tablet_41983.check_db_var('rpl_semi_sync_slave_enabled', 'ON')
+    tablet_41983.check_db_status('rpl_semi_sync_slave_status', 'OFF')
+    s = tablet_41983.mquery('', 'show slave status')
+    self.assertEqual(s[0][10], 'No')
```
Comment (pivanof) [inline test/reparent.py]: Can you please add a comment to explain what these 10 and 11 mean?

File: go/vt/tabletmanager/rpc_replication.go (hunk 13 above)
Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: No, this is topotools.ChangeType, just changes the topology.

File: test/reparent.py (hunk 14 above)
Comment (alainjobart) [inline test/reparent.py]: Done.

File: test/reparent.py (hunk 15 above)
Comment (alainjobart) [inline test/reparent.py]: Done.

File: go/vt/tabletmanager/rpc_replication.go (hunk 13 above)
Comment (pivanof) [inline go/vt/tabletmanager/rpc_replication.go]: Hm... How it worked without this call to ChangeType before then and why do we need to call it now?

Comment (alainjobart) [inline go/vt/tabletmanager/rpc_replication.go]: It didn't work before, or more accurately it left the other master as tablet type master, but fixed its replication to replicate from the new master. Not consistent.
